File: samples/python/messaging/direct_message.py
from typing import TypedDict
from langchain_anthropic import ChatAnthropic
from langgraph.graph import StateGraph, END

llm = ChatAnthropic(model="claude-haiku-4-5-20251001")


class State(TypedDict):
    question: str
    answer: str


def agent_a(state: State) -> State:
    """Agent A generates a trivia question."""
    prompt = "Generate one short trivia question about space exploration. Just the question, no answer."
    question = llm.invoke(prompt).content
    print(f"[Agent A] Question: {question}")
    return {"question": question, "answer": ""}


def agent_b(state: State) -> State:
    """Agent B receives the question and answers it."""
    prompt = f"Answer this trivia question briefly (one sentence): {state['question']}"
    answer = llm.invoke(prompt).content
    print(f"[Agent B] Answer: {answer}")
    return {"answer": answer}


def main():
    print("=== Direct Message Pattern ===\n")

    graph = StateGraph(State)
    graph.add_node("agent_a", agent_a)
    graph.add_node("agent_b", agent_b)
    graph.set_entry_point("agent_a")
    graph.add_edge("agent_a", "agent_b")
    graph.add_edge("agent_b", END)

    app = graph.compile()
    result = app.invoke({"question": "", "answer": ""})

    print("\n--- Final State ---")
    print(f"Question : {result['question']}")
    print(f"Answer   : {result['answer']}")


if __name__ == "__main__":
    main()

File: samples/typescript/messaging/direct_message.ts
import { ChatAnthropic } from "@langchain/anthropic";
import { StateGraph, END, START, Annotation } from "@langchain/langgraph";

const llm = new ChatAnthropic({ model: "claude-haiku-4-5-20251001" });

const StateAnnotation = Annotation.Root({
  question: Annotation<string>(),
  answer: Annotation<string>(),
});

type State = typeof StateAnnotation.State;

async function agentA(state: State): Promise<Partial<State>> {
  const res = await llm.invoke("Generate a short, interesting trivia question.");
  return { question: res.content as string };
}

async function agentB(state: State): Promise<Partial<State>> {
  const res = await llm.invoke(`Answer this question briefly: ${state.question}`);
  return { answer: res.content as string };
}

async function main() {
  const graph = new StateGraph(StateAnnotation)
    .addNode("agentA", agentA)
    .addNode("agentB", agentB)
    .addEdge(START, "agentA")
    .addEdge("agentA", "agentB")
    .addEdge("agentB", END)
    .compile();

  const result = await graph.invoke({ question: "", answer: "" });
  console.log("Question:", result.question);
  console.log("Answer:", result.answer);
}

main().catch(console.error);
